File: Practice/AllDomains/CoreCS/Algorithms/Strings/CommonChild.cs
/*	author:@shivkrthakur  */
using System;
using System.Collections.Generic;
using System.IO;
class Solution 
{
    static void Main(String[] args) 
	{
        /* Enter your code here. Read input from STDIN. Print output to STDOUT. Your class should be named Solution */
        string s1 = Console.ReadLine().Trim();
        string s2 = Console.ReadLine().Trim();
        int[,] matrix = new int[s1.Length+1, s2.Length+1];
        int count = FindLongestCommonSubsequence(s1, s2, s1.Length, s2.Length, matrix);
        Console.WriteLine(count);
    }
    
    static int FindLongestCommonSubsequence(string s1, string s2, int n, int m, int [,] matrix)
    {
        if(n == 0 || m == 0) return 0;
        int result = 0;
        if(matrix[n,m] > 0) result = matrix[n,m];
        else
        {
            if(s1[n-1] == s2[m-1]) result = 1 + FindLongestCommonSubsequence(s1, s2, n - 1, m - 1, matrix);
            else result = Math.Max(FindLongestCommonSubsequence(s1, s2, n , m - 1, matrix), FindLongestCommonSubsequence(s1, s2, n - 1, m, matrix));
            matrix[n,m] = result;
        }
        return result;
    }
}

File: Practice/AllDomains/CoreCS/Algorithms/Strings/CommonChild.js
/*	author:@shivkrthakur  */
process.stdin.resume();
process.stdin.setEncoding('ascii');

var input_stdin = "";
var input_stdin_array = "";
var input_currentline = 0;

process.stdin.on('data', function (data) {
    input_stdin += data;
});

process.stdin.on('end', function () {
    input_stdin_array = input_stdin.split("\n");
    main();    
});

function readLine() {
    return input_stdin_array[input_currentline++];
}

/////////////// ignore above this line ////////////////////

function commonChild(str1, str2){
   var len = str1.length;
   var arr = new Array(len + 1);
   for(var i = 0; i < len + 1; i++){
       arr[i] = new Array(len + 1);
       for(var j = 0; j < len + 1; j++)
       {
           arr[i][j] = 0;
       }
   }

   for(var i = 1; i <= len; i++){
        for(var j = 1; j <= len; j++){
            if(str1[i-1] == str2[j-1]) arr[i][j] = arr[i - 1][j - 1] + 1;
            else arr[i][j] = Math.max(arr[i][j - 1], arr[i - 1][j]);
        }
    }
    return arr[len][len];
	
	/*
	//FAILS : solution.js:45
	//function longestCommonSubsequence(str1, str2, len1, len2, arr){
	//RangeError: Maximum call stack size exceeded
    return longestCommonSubsequence(str1, str2, len - 1, len - 1, arr);  
	*/
};

function longestCommonSubsequence(str1, str2, len1, len2, arr){
    if(len1 < 0 || len2 < 0) return 0;
    if(!arr[len1][len2]){
        if(str1[len1] == str2[len2]) arr[len1][len2] = 1 + longestCommonSubsequence(str1, str2, len1 - 1, len2 - 1, arr);
        else arr[len1][len2] = Math.max(longestCommonSubsequence(str1, str2, len1, len2 - 1, arr), longestCommonSubsequence(str1, str2, len1 - 1, len2, arr));
    }
    return arr[len1][len2];
}

function main() {
    var s1 = readLine();
    var s2 = readLine();
    var result = commonChild(s1, s2);
    process.stdout.write("" + result + "\n");

}
